subcommand/consul-sidecar/command_test.go

```diff
@@ -194,42 +191,50 @@ func TestMergedMetricsServer(t *testing.T) {
 
 	for _, c := range cases {
 		t.Run(c.name, func(t *testing.T) {
+			randomPorts := freeport.MustTake(2)
 			ui := cli.NewMockUi()
 			cmd := Command{
 				UI:                       ui,
 				flagEnableMetricsMerging: true,
-				flagMergedMetricsPort:    "20100",
-				flagServiceMetricsPort:   "8080",
+				flagMergedMetricsPort:    fmt.Sprint(randomPorts[0]),
+				flagServiceMetricsPort:   fmt.Sprint(randomPorts[1]),
 				flagServiceMetricsPath:   "/metrics",
 				logger:                   hclog.Default(),
 			}
 
+			server := cmd.createMergedMetricsServer()
+
+			// Override the cmd's envoyMetricsGetter and serviceMetricsGetter
+			// with stubs.
+			em := &envoyMetrics{}
+			sm := &serviceMetrics{}
 			if c.runEnvoyMetricsServer {
-				mockEnvoyMetricsServer := newMockEnvoyMetricsServer(t)
-				mockEnvoyMetricsServer.Start()
-				defer mockEnvoyMetricsServer.Close()
+				cmd.envoyMetricsGetter = em
 			}
 			if c.runServiceMetricsServer {
-				mockServiceMetricsServer := newMockServiceMetricsServer(t)
-				mockServiceMetricsServer.Start()
-				defer mockServiceMetricsServer.Close()
+				cmd.serviceMetricsGetter = sm
 			}
 
-			server := cmd.createMergedMetricsServer()
 			go func() {
 				_ = server.ListenAndServe()
 			}()
 			defer server.Close()
 
-			// Wait for the servers to come up
+			// Wait for the merged metrics server to come up
 			time.Sleep(2 * time.Second)
```

Comment:
i know this was not added as a part of the refactor, but is there a way to test this without the sleep? Can we run the asserts in a retry.Run() bloack? A hard wait of 2 seconds per test case feel kinda high.

Reply:
There are some tests that use that.

Reply:
Yeah this is a great suggestion, I made changes to both tests that wait for the server to come up to use retry instead!